### PR TITLE
[SYCLomatic] Ensure wrappers are only generated under C language linkage and for __global__ functions

### PR DESCRIPTION
**Pull request by jzc**
Signed-off-by: Cai, Justin <justin.cai@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 1]
```diff
@@ -2459,6 +2459,9 @@ void DeviceFunctionDeclInModule::buildParameterInfo(const FunctionDecl *FD) {
 }
 
 void DeviceFunctionDeclInModule::buildWrapperInfo(const FunctionDecl *FD) {
+  GenerateWrapper =
```
Comment (0x12CC) [inline clang/lib/DPCT/AnalysisInfo.cpp]: This condition looks correct but can it be in the definition of `isModuleFunction`? I think that moving it there would eliminate the need for a new data member.

Comment (jzc) [reply clang/lib/DPCT/AnalysisInfo.cpp]: I think that makes sense, I will implement that.

**Review by 0x12CC — APPROVED** [review]

File: clang/test/dpct/module_wrapper_gen.cu [hunk 2]
```diff
@@ -0,0 +1,26 @@
+// RUN: dpct --extra-arg="--ptx" \
+// RUN:      --out-root=%T/module_wrapper_gen \
+// RUN:      --cuda-include-path="%cuda-path/include" %s
+// RUN: FileCheck %s --input-file=%T/module_wrapper_gen/module_wrapper_gen.dp.cpp
+
+// START
+
```
Comment (LU-JOHN) [inline clang/test/dpct/module_wrapper_gen.cu:7]: Can we add a functions:

__device__
void int_test() {  // check non-extern-C device non-operator is not wrapped
}

__global__
void int_wow() { // check non-extern-C global is not wrapped
}

to ensure that these are not wrapped.

Comment (jzc) [reply clang/test/dpct/module_wrapper_gen.cu:7]: I've added these cases in the test

File: clang/test/dpct/module_wrapper_gen.cu [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+__device__ float2 operator+(float2 a, float2 b) {
+  return float2{a.x - b.x, a.y - b.y};
+}
+
+extern "C" __device__ void externCNonKernel() {}
+
+__device__ void deviceFun() {}
+
+__global__ void nonExternCKernel() {}
```
Comment (ziranzha) [inline clang/test/dpct/module_wrapper_gen.cu:16]: move to line:23, or it is not checked.

**Review by LU-JOHN — APPROVED** [review]
LGTM

File: clang/test/dpct/module_wrapper_gen.cu (hunk 3 above)
Comment (jzc) [reply clang/test/dpct/module_wrapper_gen.cu:16]: I'm not sure that affects the behavior of the test -- the test should pass if there is exactly one `_wrapper` between the `// START` and `// END` comments, and moving this to line 23 should not affect that.

**Pull request merged**